### PR TITLE
In process task scanner

### PR DESCRIPTION
**Pull request by tintoy**
Implements tintoy/msbuild-project-tools-server#42.

@DoctorKrolic - are you able to run from this branch and see if it works for you? If not, I'll build a VSIX locally and attach it to the ticket.

### CODE REVIEW

**Comment by DoctorKrolic** [comment]
I tried to use language server from this branch in VS Code extension and I get the same exception as in the CI (except a bit different path to dll, but this is not the point) when I try to trigger completion of tasks. I investigated it a little bit and here are my results:
- [Here](https://github.com/tintoy/msbuild-project-tools-server/blob/65a48e810ed213b1c7b308308d1f304a5ff3100b/src/LanguageServer.SemanticModel.MSBuild/MSBuildTaskScanner.cs#L140) we load types from assembly, this works correct.
- Then we [filter these types](https://github.com/tintoy/msbuild-project-tools-server/blob/65a48e810ed213b1c7b308308d1f304a5ff3100b/src/LanguageServer.SemanticModel.MSBuild/MSBuildTaskScanner.cs#L147-L153), which is the root cause of a problem. I tried to set a breakpoint inside the filter and it works correctly until it faces 1 specific type, which throws during `type.IsClass` check. I guess, we can wrap all "dangerous" checks in a try-catch and just filter type out if it throws. But you might find better solution for this, I am not a great expert in reflection area

**Comment by tintoy** [comment]
Thanks! Will do some inspection and see what I can find :)

**Comment by DoctorKrolic** [comment]
No worries, this is definitely not critical priority

File: MSBuildProjectTools.sln [hunk 1]
```diff
@@ -33,8 +33,6 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "LanguageServer.SemanticMode
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "LanguageServer.Common", "src\LanguageServer.Common\LanguageServer.Common.csproj", "{5AD7AEFF-4E5C-474C-8BD6-DEBD354A803C}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "LanguageServer.TaskReflection", "src\LanguageServer.TaskReflection\LanguageServer.TaskReflection.csproj", "{C594F3C8-8E90-4AE2-B75C-DB17AC652735}"
-EndProject
```
Comment (DoctorKrolic) [inline MSBuildProjectTools.sln:37]: It seems like you removed project reference from the solution, but didn't delete actual project files. Remove this files before merging pls

File: global.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "sdk": {
-    "version": "6.0.100",
+    "version": "6.0.203",
```
Comment (DoctorKrolic) [inline global.json:3]:
```suggestion
    "version": "6.0.413",
```

File: src/LanguageServer.Engine/Documents/ProjectDocument.cs [hunk 3]
```diff
@@ -741,25 +741,24 @@ public Range GetRange(ExpressionNode expression, Position relativeToPosition)
...
 
+            DotNetRuntimeInfo currentRuntime = DotNetRuntimeInfo.GetCurrent();
+
             List<string> taskAssemblyFiles = new List<string>
             {
                 // Include "built-in" tasks.
-                Path.Combine(DotNetRuntimeInfo.GetCurrent().BaseDirectory, "Microsoft.Build.Tasks.Core.dll"),
-                Path.Combine(DotNetRuntimeInfo.GetCurrent().BaseDirectory, "Roslyn", "Microsoft.Build.Tasks.CodeAnalysis.dll")
+                Path.Combine(currentRuntime.BaseDirectory, "Microsoft.Build.Tasks.Core.dll"),
+                Path.Combine(currentRuntime.BaseDirectory, "Roslyn", "Microsoft.Build.Tasks.CodeAnalysis.dll")
```
Comment (DoctorKrolic) [inline src/LanguageServer.Engine/Documents/ProjectDocument.cs:761]: Nice

File: src/LanguageServer.SemanticModel.MSBuild/MSBuildTaskMetadataCache.cs [hunk 4]
```diff
@@ -24,10 +27,19 @@ public sealed class MSBuildTaskMetadataCache
...
+        /// </param>
+        public MSBuildTaskMetadataCache(ILogger logger)
         {
+            Log = logger;
         }
 
+        /// <summary>
+        ///     An optional logger to use for type-reflection diagnostics.
+        /// </summary>
+        ILogger Log { get; }
```
Comment (DoctorKrolic) [inline src/LanguageServer.SemanticModel.MSBuild/MSBuildTaskMetadataCache.cs:41]: I've seen this multiple times in the codebase, but I think it's better to have private elements as actual fields rather than private properties. Makes clearer separation between the two. Better to be done in a separate PR, so just want your opinion about this

Comment (tintoy) [reply src/LanguageServer.SemanticModel.MSBuild/MSBuildTaskMetadataCache.cs:41]: Hmm - yeah fair enough; this is another one of those deliberate choices and I get that it may seem somewhat idiosyncratic but I've just woken up so I'll try to articulate the reasoning behind it later today.

(FWIW I was thinking just yesterday about places where I've done this in the codebase at work and why so I'm going to have to remember why either way; "logging is a special case" doesn't really suffice as an explanation)

**Comment by tintoy** [comment]
Thanks for taking the time to review BTW!

**Comment by DoctorKrolic** [comment]
@tintoy Any updates on this?

**Comment by tintoy** [comment]
Sorry, forgot about this one; will try to finish it off today 🙂

File: src/LanguageServer.SemanticModel.MSBuild/MSBuildTaskMetadataCache.cs (hunk 4 above)
Comment (tintoy) [reply src/LanguageServer.SemanticModel.MSBuild/MSBuildTaskMetadataCache.cs:41]: If you don't mind, I'm going to create an issue to discuss this one, as I don't want it to hold up this PR (since it's a more global change that would be needed to keep things consistent).

Comment (tintoy) [reply src/LanguageServer.SemanticModel.MSBuild/MSBuildTaskMetadataCache.cs:41]: Moved to tintoy/msbuild-project-tools-server#61
